### PR TITLE
Add the context menu

### LINKED ISSUE #6980: Add a right-click context menu to DCA
As discussed in the Contao call, we could add a right-click context menu that has additional buttons such as "new into" and "new after". It could also have options such as "add new child elements" similar to what Slack does:

_Originally posted by @leofeyer in https://github.com/contao/contao/issues/6790#issuecomment-1983837872_

### PR DESCRIPTION
**Pull request by aschempp**
Fixes https://github.com/contao/contao/issues/6980

This simplifies the operations, and groups the lesser-used ones into a menu button.

Bildschirmfoto 2024-12-20 um 14 54 48
Bildschirmfoto 2024-12-20 um 14 54 56

The same menu can also be reached by right-clicking anywhere on the item (context menu):
Bildschirmfoto 2024-12-20 um 14 55 31

Having the menu button (over just a context menu) is important for mobile and also for accessibility.

### CODE REVIEW

**Comment by Toflar** [comment]
As far as I remember we said this behavior should be configurable because some users preferred the "more icons but one less click"-way? Or am I remembering this wrong?

**Comment by zoglo** [comment]
> As far as I remember we said this behavior should be configurable because some users preferred the "more icons but one less click"-way? Or am I remembering this wrong?

We wanted to move it to the user settings afaik

File: core-bundle/contao/themes/flexible/styles/main.css [hunk 1]
```diff
@@ -1516,7 +1516,15 @@ table.tl_listing {
...
 }
 
-.tl_right button, .tl_content_right button {
+.tl_right > nav > ul,
+.tl_right_nowrap > nav > ul,
+.tl_content_right > nav > ul {
+	display: flex;
+	justify-content: flex-end;
+	gap: 3px;
+}
```
Comment (zoglo) [inline core-bundle/contao/themes/flexible/styles/main.css]: This might pose some problems such as icon alignment within the new nav / ul / li structure.

See files for example:
image

could be fixed by omitting the `display: flex` completely and using something like this:
```css
:is(.tl_content_right, .tl_right, .tl_right_nowrap) > nav > ul > li {
    display: inline-block;
}

/* vertically align the context-menu button with the image actions */
.operations-menu-container > button {
    vertical-align: middle;
}
```
image

File: core-bundle/contao/templates/twig/backend/data_container/operations.html.twig [hunk 2]
```diff
@@ -1,20 +1,43 @@
...
+                {% set has_submenu = true %}
+            {% else %}
+                {{ _self.operation(operation) }}
+            {% endif %}
+        {% endfor %}
+        {% if has_submenu or not has_primary %}
+            <li class="operations-menu-container" data-contao--operations-menu-target="container">
+                {% if has_primary %}
+                    <button title="More" data-contao--operations-menu-target="controller">
+                        <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-ellipsis"><circle cx="12" cy="12" r="1"/><circle cx="19" cy="12" r="1"/><circle cx="5" cy="12" r="1"/></svg>
```
Comment (zoglo) [inline core-bundle/contao/templates/twig/backend/data_container/operations.html.twig]: Gonna quote someone in advance 🐆 
> https://github.com/contao/contao/pull/7675#discussion_r1891894717 If this wasn't an inline SVG, Webpack would optimize it automatically it and we could provide a dark mode variant.

File: core-bundle/contao/templates/twig/backend/data_container/operations.html.twig [hunk 3]
```diff
@@ -1,20 +1,43 @@
-{% for operation in operations %}
-    {{ _self.operation(operation) }}
-{% endfor %}
+{% set has_submenu = false %}
+<nav>
```
Comment (zoglo) [inline core-bundle/contao/templates/twig/backend/data_container/operations.html.twig]: _Might be an accessibility concern_
I am unsure of the usage of `nav` here since it's usually used for major blocks of navigation links.
This would create lots of navigations on every page when there's many context-menus being rendered.

File: package.json [hunk 4]
```diff
@@ -19,5 +19,8 @@
         "build": "encore production --progress"
     },
     "private": true,
-    "packageManager": "npm@10.8.2"
+    "packageManager": "npm@10.8.2",
+    "dependencies": {
+        "accessible-menu": "^4.1.2"
+    }
```
Comment (zoglo) [inline package.json]: Imho a devDependency since it's being built either way

**Review by zoglo — CHANGES_REQUESTED** [review]
There are a few things that could / might need changes.

There's also overflow and z-index issues within the parent view which may not be as easy to fix within the stacking-context.

1. Anything with an overflow: hidden on the container may not work
image

2. Since the following element has a z-index of 1, it'll be higher in the stacking-context:
image

3. We actually have two navigations here when opening one and then right-clicking
image


### Possible solution

Would we be able to open the second content-menu on button click instead of the one in the dom? That will most likely solve all the issues mentioned above:
image

File: package.json (hunk 4 above)
Comment (aschempp) [reply package.json]: Everything is built. A dev dependency to me is stuff that is needed for the build process, but not stuff that is actually used in the "front end".

File: core-bundle/contao/templates/twig/backend/data_container/operations.html.twig (hunk 2 above)
Comment (leofeyer) [reply core-bundle/contao/templates/twig/backend/data_container/operations.html.twig]: Yes, that needs to be an icon indeed.

File: core-bundle/contao/templates/twig/backend/data_container/operations.html.twig [hunk 5]
```diff
@@ -1,20 +1,43 @@
...
+            {% if has_primary and not operation.primary|default(false) %}
+                {% set has_submenu = true %}
+            {% else %}
+                {{ _self.operation(operation) }}
+            {% endif %}
+        {% endfor %}
+        {% if has_submenu or not has_primary %}
+            <li class="operations-menu-container" data-contao--operations-menu-target="container">
+                {% if has_primary %}
+                    <button title="More" data-contao--operations-menu-target="controller">
```
Comment (leofeyer) [inline core-bundle/contao/templates/twig/backend/data_container/operations.html.twig]: `title="More"` ?

File: core-bundle/contao/themes/flexible/styles/main.css (hunk 1 above)
Comment (leofeyer) [reply core-bundle/contao/themes/flexible/styles/main.css]: Yes, this needs to be fixed.

File: core-bundle/tests/EventListener/DataContainer/DefaultOperationsListenerTest.php [hunk 6]
```diff
@@ -409,6 +409,38 @@ public function testDoesNotAddDefaultsIfOneOperationHasADefaultKey(): void
...
+    {
+        /** @phpstan-var array $GLOBALS (signals PHPStan that the array shape may change) */
+        $GLOBALS['TL_DCA']['tl_foo'] = [
+            'list' => [
+                'sorting' => [
+                    'mode' => DataContainer::MODE_SORTED,
+                ],
+                'operations' => [
+                    '-edit',
+                    '+show',
```
Comment (leofeyer) [inline core-bundle/tests/EventListener/DataContainer/DefaultOperationsListenerTest.php]: Is this a new feature? How does it work?

File: core-bundle/contao/templates/twig/backend/data_container/operations.html.twig (hunk 3 above)
Comment (aschempp) [reply core-bundle/contao/templates/twig/backend/data_container/operations.html.twig]: Good point. Changed in 03c2919c06c426da9d7aa37b8fd2511d3a07a2fe

File: core-bundle/contao/templates/twig/backend/data_container/operations.html.twig (hunk 2 above)
Comment (aschempp) [reply core-bundle/contao/templates/twig/backend/data_container/operations.html.twig]: I think the icon works very well in both modes. But I fixed a dark mode issue in a0ddeba1008ba40f1b278c3ffa7a0a8c7212bbb5

File: core-bundle/tests/EventListener/DataContainer/DefaultOperationsListenerTest.php (hunk 6 above)
Comment (aschempp) [reply core-bundle/tests/EventListener/DataContainer/DefaultOperationsListenerTest.php]: Yes it is. You somehow need a way to mark a string-only operation as primary. I changed it in 23a7b86c0f45a66d3b50302a4ae3588bb65e8a48 though (reverted to the first draft actually). You now would use `!show` to mark an operation as primary. String operations are never _primary_ by default.

File: core-bundle/contao/templates/twig/backend/data_container/operations.html.twig (hunk 2 above)
Comment (zoglo) [reply core-bundle/contao/templates/twig/backend/data_container/operations.html.twig]: Should still be an icon as it'll bloat the HTML on 100+ pages open (no need to render the same svg many times, just have it preload once in an URL 😄)

File: core-bundle/contao/themes/flexible/styles/main.css (hunk 1 above)
Comment (aschempp) [reply core-bundle/contao/themes/flexible/styles/main.css]: should all be fixed by d0c9057aac438624a47bca783fea63202c261c19

File: core-bundle/contao/templates/twig/backend/data_container/operations.html.twig (hunk 5 above)
Comment (aschempp) [reply core-bundle/contao/templates/twig/backend/data_container/operations.html.twig]: Fixed in d0c9057aac438624a47bca783fea63202c261c19

File: core-bundle/contao/themes/flexible/styles/main.css [hunk 7]
```diff
@@ -1529,6 +1528,62 @@ table.tl_listing {
 	margin-left: 1px;
 }
 
+.operations > ul {
+	display: flex;
+	justify-content: flex-end;
+	align-items: center;
+	gap: 3px;
+	flex-wrap: wrap;
+	list-style: none;
```
Comment (zoglo) [inline core-bundle/contao/themes/flexible/styles/main.css:1562]:
```suggestion
	list-style: none;
	margin: 0;
```

We would want to always disable the browser styles for this offset to not happen :)
image

File: core-bundle/contao/themes/flexible/styles/main.css [hunk 8]
```diff
@@ -1529,6 +1528,62 @@ table.tl_listing {
...
+.operations-menu .virtual-title a:after {
+	content: attr(title);
+	padding-left: 5px;
+}
+
+.operations-menu .virtual-title a[data-original-title]:after {
+	content: attr(data-original-title);
+	padding-left: 5px;
+}
+
```
Comment (zoglo) [inline core-bundle/contao/themes/flexible/styles/main.css:1613]:
```suggestion
}

.tl_content_right:has(.operations-menu.show) {
	z-index: 2;
}
```

What would we do without the `:has` selector :D

image

**Review by zoglo — CHANGES_REQUESTED** [review]
I added some `changes` that will fix a few bugs @aschempp.

On the UX side, I've seen two things that can be fixed:

1. Overflow on mobile due to the position: fixed and `visibility: hidden` which **is needed** for the positioning to properly work. We can bypass this (and many other overflow-issues) by adding a new stacking-context to `#main`:

```diff
/* Tablet */
@media (max-width: 991px) {
	#container {
		display: block;
	}

	#main, #left {
		float: none;
	}

	#main {
		width: 100% !important;
		position: relative;
		transition: transform .2s ease;
		-webkit-transform: none;
-		transform: none;
+		transform: translate(0,0,0);
		will-change: transform;
+		overflow: hidden;
		
	}
```
image


2. The width gets ultra long on files but as discussed in the dev-call, we don't really need to show the long name anymore and go with simple translations.
image

_3. Using a svg-image rather than the inline-svg is better since it won't bloat the HTML as much but it's not an issue to me._

___

I'm fine if the requested changes within the code are merged + the stacking context on `#main`.

File: core-bundle/contao/themes/flexible/styles/main.css [hunk 9]
```diff
@@ -1529,6 +1528,62 @@ table.tl_listing {
 	margin-left: 1px;
 }
 
+.operations > ul {
+	display: flex;
+	justify-content: flex-end;
+	align-items: center;
+	gap: 3px;
+	flex-wrap: wrap;
```
Comment (zoglo) [inline core-bundle/contao/themes/flexible/styles/main.css]:
```suggestion
```
Removing that for consistency on mobile might be better:
image

**Review by zoglo — CHANGES_REQUESTED** [review]

File: core-bundle/contao/themes/flexible/styles/main.css (hunk 8 above)
Comment (aschempp) [reply core-bundle/contao/themes/flexible/styles/main.css:1613]: There is already a `z-index:10` on `.operations-menu`. I guess we need to `!important` that instead of just using a longer selector to override?

File: core-bundle/contao/themes/flexible/styles/main.css (hunk 9 above)
Comment (aschempp) [reply core-bundle/contao/themes/flexible/styles/main.css]: I added this because the operations previously were inline and would wrap as well, or am I wrong?

Comment (zoglo) [reply core-bundle/contao/themes/flexible/styles/main.css]: They wouldn't wrap before
image

File: core-bundle/contao/themes/flexible/styles/main.css (hunk 8 above)
Comment (zoglo) [reply core-bundle/contao/themes/flexible/styles/main.css:1613]: Nope, same issue as always. The stacking context matters. The next sibling of `tl_right` has a the same `z-index`, thus higher stacking-context so `z-index: 1` of `tl_right` would completely nullify the `z-index: 10`.

TL;DR: It just has a z-index of 10 within it's parent that has a z-index of 1... not in the whole document :)

**Comment by zoglo** [comment]
We also need to remove the `will-change: transform` on #main for the offset to properly translate on mobile.

File: core-bundle/contao/themes/flexible/styles/main.css (hunk 9 above)
Comment (aschempp) [reply core-bundle/contao/themes/flexible/styles/main.css]: Removed for now in 9362b2ce48c092bb296318ae7070b898976c6157. It was however possible to wrap operations previously (like Isotope did). Need to see how we can still fix that :/

File: core-bundle/contao/themes/flexible/styles/main.css (hunk 8 above)
Comment (aschempp) [reply core-bundle/contao/themes/flexible/styles/main.css:1613]: that has become irrelevant with `position:fixed`, right?

Comment (zoglo) [reply core-bundle/contao/themes/flexible/styles/main.css:1613]: > that has become irrelevant with `position:fixed`, right?

Nope, you can test it in any content-element right now when opening the context-menu with the 3 dots:
image

**Review by leofeyer — CHANGES_REQUESTED** [review]
There are still some issues that need to be resolved:

### Right click on icon

I can no longer right click on the icons now without the context menu opening up. The context menu should only open if I

- click the "more" icon or
- right click on the row

but **not** if I right click on an icon or a link inside the row.

### CSS issues

**Comment by zoglo** [comment]
> ### CSS issues

Should be fixed with that: https://github.com/contao/contao/pull/7816#discussion_r1913553741

**Review by zoglo — APPROVED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.

**Comment by timgatzky** [comment]
> As far as I remember we said this behavior should be configurable because some users preferred the "more icons but one less click"-way? Or am I remembering this wrong?

-> is this applied already? I cannot find it in the user profile.

In general:
Hiding important interactive elements is always a bad thing. (Especially for new user and even for returning users). 
-> What makes a button important enough to not be hidden by default? 

When opened (by accident or wrong decision) the box overlaps the other box-buttons. So the user must close the open onet first (one click or reload), then find the one that is wanted -> another click to open -> find the operation wanted and another click to finally fire it. 
This is a ton more clicking and rethinking where things are and might get frustrating very fast.

A collapsed context menu might be a solution for small screens to avoid line breaks or overlaps, but when there is enough space for the buttons, operations should be accessible at any time without unnecessary more clicking. 

This feature should not be default. It makes sense for small screens only or the user enables it.

**Comment by leofeyer** [comment]
This pull request is closed. Please create a new issue instead.
